### PR TITLE
Reimplement LogManager sysmodule

### PR DESCRIPTION
**Pull request by XorTroll**
Reimplements LogManager (LogManager.Prod in retail) sysmodule, redirecting log data to log files in the SD card.
There isn't much left to do (everything seems to work fine), it just needs testing and a proper review.

### CODE REVIEW

**Comment by XorTroll** [comment]
So, one of the aspects I have in my mind right now is the idea of specifying log verbosity via an atmosphere setting, since these logs have many fields which could be logged, but logging all of them (what this reimplementation currently does) might be too verbose for users/devs which just want the text log and don't care about the rest of the fields.

**Comment by misson20000** [comment]
I'd rather see logs stores in a machine-readable format for later searching and filtering.

**Comment by XorTroll** [comment]
This last commit changed many things:
- Now logs are saved in binary format (agree, it's better to save them this way and then parse/print them elsewhere). Save format: `sdmc:/atmosphere/debug_logs/<hex-program-id>/<hex-log-id>.bin`, where the log ID is the current system tick.
- Logs are removed everytime the sysmodule starts (every reboot), since games like Rocket League generate tons of logs which end up wasting a lot of space (in this case, RL generates hundreds of logs just before the title screen is loaded).
- As a way to let other sysmodules/processes get advantage of our logging, added two extra commands to lm service: one to get an event which signals when any process logs, and another one to get the last log's information (log ID and program ID), so that the sysmodule can read the logged data from SD.
- Removed all log packet decoding code, since it is no longer used in LogManager.
As a sidenote, I've tested the extra commands with a simple TCP logger between PC and a sysmodule, and I managed to print everything games/processes log in real time on PC without any issues.

File: stratosphere/LogManager/LogManager.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
+{
+    "name": "LogManager",
+    "title_id": "0x0100000000000015",
+    "title_id_range_min": "0x0100000000000015",
+    "title_id_range_max": "0x0100000000000015",
+    "main_thread_stack_size": "0x00004000",
+    "main_thread_priority": 10,
```
Comment (SciresM) [inline stratosphere/LogManager/LogManager.json:7]: This is incorrect. LogManager.MainThread has a system thread priority of 10; this corresponds to an actual hos priority of 38.

File: stratosphere/LogManager/LogManager.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            "type": "min_kernel_version",
+            "value": "0x0030"
+        },
+        {
+            "type": "debug_flags",
+            "value": {
+                "allow_debug": false,
+                "force_debug": true
+            }
```
Comment (SciresM) [inline stratosphere/LogManager/LogManager.json:100]: Should there be a handle table descriptor to match official LogManager?

File: stratosphere/LogManager/source/impl/lm_logging.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
+#include "lm_logging.hpp"
```
Comment (SciresM) [inline stratosphere/LogManager/source/impl/lm_logging.cpp:1]: Atmosphere copyright header + #include <stratosphere.hpp>

File: stratosphere/LogManager/source/impl/lm_logging.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+    }
+
+    void SetCanAccessFs(bool can_access) {
+        std::scoped_lock lk(g_logging_lock);
+        g_can_access_fs = can_access;
+    }
+
+    Result WriteLogPackets(std::vector<LogPacketBuffer> &packet_list) {
```
Comment (SciresM) [inline stratosphere/LogManager/source/impl/lm_logging.cpp:34]: std::vector --> no.

File: stratosphere/LogManager/source/lm_service.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            if (settings::fwdbg::GetSettingsItemValue(&save_debug_logs, sizeof(save_debug_logs), "atmosphere", "logmanager_save_debug_logs") != sizeof(save_debug_logs)) {
+                return false;
+            }
+
+            return save_debug_logs != 0;
+        }
+
+    }
+
+    void Logger::Log(const sf::InAutoSelectBuffer &buf) {
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_service.cpp:19]: This doesn't seem to match the code RE I've done on lm in the past...

File: stratosphere/LogManager/source/lm_types.hpp [hunk 6]
```diff
@@ -0,0 +1,14 @@
+
+#pragma once
+#include <stratosphere.hpp>
+
+namespace ams::lm {
+
+    enum class LogDestination : u32 {
+        TMA = 1,
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_types.hpp:8]: Atmosphere uses (1 << ...) for bitflag enums.

File: stratosphere/Makefile [hunk 7]
```diff
@@ -1,4 +1,4 @@
-MODULES := loader ncm pm sm boot ams_mitm spl eclct.stub ro creport fatal dmnt boot2 erpt pgl jpegdec
+MODULES := loader ncm pm sm boot ams_mitm spl eclct.stub ro creport fatal dmnt boot2 erpt pgl jpegdec LogManager
```
Comment (SciresM) [inline stratosphere/Makefile:1]: module should really be called lm (and just have LogManager as npdm name)

File: stratosphere/LogManager/source/lm_service.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
+
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_service.hpp:1]: Atmosphere copyright header, #include <stratosphere.hpp>

File: stratosphere/LogManager/source/lm_service.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#include "lm_service.hpp"
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_service.cpp:1]: Atmosphere copyright header, #include <stratosphere.hpp>

File: stratosphere/LogManager/source/lm_main.cpp [hunk 10]
```diff
@@ -0,0 +1,150 @@
+#include "lm_service.hpp"
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_main.cpp:1]: Atmosphere copyright header, #include <stratosphere.hpp>

File: stratosphere/LogManager/source/impl/lm_logging.hpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
+
```
Comment (SciresM) [inline stratosphere/LogManager/source/impl/lm_logging.hpp:1]: Atmosphere copyright header, #include <stratosphere.hpp>

File: stratosphere/LogManager/source/lm_types.hpp [hunk 12]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_types.hpp:1]: Atmosphere copyright header.

File: stratosphere/LogManager/source/lm_service.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+    void Logger::SetDestination(LogDestination destination) {
+        /* TODO: shall we make use of this value? */
+        this->destination = destination;
+    }
+
+    void LogService::OpenLogger(const sf::ClientProcessId &client_pid, sf::Out<std::shared_ptr<Logger>> out_logger) {
+        u64 program_id = 0;
+        /* Apparently lm succeeds on many/all commands, so we will succeed on them too. */
+        pminfoGetProgramId(&program_id, static_cast<u64>(client_pid.GetValue()));
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_service.cpp:62]: This should use the libstratosphere  wrapper.

File: stratosphere/LogManager/source/lm_main.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    /* Cleanup services. */
+    setsysExit();
+    pscmExit();
+    fsExit();
+    pminfoExit();
+}
+
+namespace {
+
+    /* TODO: these domain/domain object amounts work fine, but which ones does N actually use? */
```
Comment (SciresM) [inline stratosphere/LogManager/source/lm_main.cpp:73]: Yeah, you should check what N uses.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+#pragma once
+#include <stratosphere/lm/detail/lm_detail_log_types.hpp>
+
+namespace ams::lm::detail {
+
+    class LogPacketTransmitterBase {
+        NON_COPYABLE(LogPacketTransmitterBase);
+        NON_MOVEABLE(LogPacketTransmitterBase);
+        public:
+            using FlushFunction = bool (*)(const void*, size_t);
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:25]: FlushFunction takes in const u8 *, not const void *.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        private:
+            LogPacketHeader *header;
+            u8 *log_buffer_start;
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:28]: Kind of verbose, consider e.g. *start; *end; *payload; *current;

This is not a required change.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+            u8 *log_buffer_end;
+            u8 *log_buffer_payload_start;
+            u8 *log_buffer_payload_current;
+            bool is_tail;
+            FlushFunction flush_function;
+        public:
+            LogPacketTransmitterBase(u8 *log_buffer, size_t log_buffer_size, FlushFunction flush_fn, diag::LogSeverity severity, bool verbosity, u64 process_id, bool head, bool tail);
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:35]: log_buffer should be void *, do not abbreviate function as fn. use "func", if you must..

Severity/verbosity are u8, not enum/bool here.

File: libraries/libstratosphere/include/stratosphere/fs/fs_sd_card.hpp [hunk 18]
```diff
@@ -18,8 +18,28 @@
...
+            Result GetEventHandle(os::SystemEvent *out_event, os::EventClearMode clear_mode);
+    };
+
     Result MountSdCard(const char *name);
 
     Result MountSdCardErrorReportDirectoryForAtmosphere(const char *name);
 
+    bool IsSdCardInserted();
+
+    Result OpenSdCardDetectionEventNotifier(SdCardDetectionEventNotifier *out);
```
[the remark this answers is not in the document]
Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/fs/fs_sd_card.hpp:43]: I will handle this.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+            inline void PushDataChunk(LogDataChunkKey key, const char *str, size_t str_len) {
+                return this->PushDataChunkImpl(key, str, str_len, true);
+            }
+
+            constexpr size_t GetRemainSize() {
+                return static_cast<size_t>(this->log_buffer_end - this->log_buffer_payload_current);
+            }
+
+            static constexpr size_t GetRequiredSizeToPushUleb128(u64 value) {
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:59]: This seems fake/has messed up access controls. I'll refactor this myself after the detail -> impl changes are made.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    };
+
+    class LogPacketTransmitter : public LogPacketTransmitterBase {
+        NON_COPYABLE(LogPacketTransmitter);
+        NON_MOVEABLE(LogPacketTransmitter);
+        public:
+            using LogPacketTransmitterBase::LogPacketTransmitterBase;
+
+            void PushLogSessionBegin() {
+                u8 dummy_value = 1;
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:76]: This is actually the boolean value true.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+        constexpr inline u32 GetPayloadSize() const {
+            return this->payload_size;
+        }
+
+    };
+    static_assert(sizeof(LogPacketHeader) == 0x18);
+
+    enum LogDataChunkKey : u8 {
+        LogDataChunkKey_LogSessionBegin    = 0,  ///< Log session begin.
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:115]: Atmosphere doesn't use doxygen, shouldn't have doxygen comments.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+namespace ams::lm::detail {
+
+    enum LogPacketFlags {
+        LogPacketFlags_Head         = (1 << 0),
+        LogPacketFlags_Tail         = (1 << 1),
+        LogPacketFlags_LittleEndian = (1 << 2)
+    };
+
+    struct BinaryLogHeader {
+        static constexpr u32 Magic = util::FourCC<'h','p','h','p'>::Code;
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:28]: Is this the correct endianness?

Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:28]: Given the code is conceptually "phph" (packet header, packet header), may well do to use reverse code with that encoding.

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        u8 version;
+        u8 pad[3];
+    };
+
+    struct LogPacketHeader {
+        u64 process_id;
+        u64 thread_id;
+        u8 flags;
+        u8 pad;
+        diag::LogSeverity severity;
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:40]: should be u8

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+        char verbosity;
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:41]: should be u8

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+        constexpr inline void SetThreadId(u64 thread_id) {
+            this->thread_id = thread_id;
+        }
+
+        constexpr inline void SetLittleEndian(bool le) {
+            if (le) {
+                this->flags |= LogPacketFlags_LittleEndian;
+            }
+            else {
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:64]: incorrect formatting for ams style

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+
+        constexpr inline bool IsHead() const {
+            return this->flags & LogPacketFlags_Head;
+        }
+
+        constexpr inline void SetHead(bool head) {
+            if (head) {
+                this->flags |= LogPacketFlags_Head;
+            }
+            else {
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:85]: incorrect formatting for ams style

File: libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+        constexpr inline void SetPayloadSize(u32 size) {
+            this->payload_size = size;
+        }
+
+        constexpr inline u32 GetPayloadSize() const {
+            return this->payload_size;
+        }
+
+    };
+    static_assert(sizeof(LogPacketHeader) == 0x18);
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:112]: static_assert(util::is_pod<LogPacketHeader>::value);

File: libraries/libstratosphere/include/stratosphere/lm/lm_api.hpp [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Comment (SciresM) [inline libraries/libstratosphere/include/stratosphere/lm/lm_api.hpp:16]: gap

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace ams::diag::detail {
+
+    namespace {
+
+        os::SdkMutex g_log_observer_lock;
+
+        constexpr const char *g_log_severity_color_table[LogSeverity_Count] = {
+            "\x1B[90m",        /* Dark gray for Trace. */
+            "",                /* No special color for Info. */
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:26]: This is nullptr in Nintendo's implementation, I believe. Should we match?

File: libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_log.hpp [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+#include <stratosphere/diag/diag_log_types.hpp>
+
+namespace ams::diag::detail {
+
+    void LogImpl(const LogMetaData &log_metadata, const char *fmt, ...);
+    void VLogImpl(const LogMetaData &log_metadata, const char *fmt, std::va_list va_args);
+    void CallAllLogObserver(const LogMetaData &log_metadata, const LogBody &log_body);
```
[the remark this answers is not in the document]
Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_log.hpp:23]: It is, but it's internal/not exposed. Why are you exposing it?

File: libraries/libstratosphere/include/stratosphere/diag/diag_log_observer.hpp [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+        bool is_registered;
+        void *user_data;
+    };
+
+    void InitializeLogObserverHolder(LogObserverHolder *observer_holder, LogObserverHolder::LogFunction log_function, void *user_data);
+
+    void RegisterLogObserver(LogObserverHolder *observer_holder);
+    void UnregisterLogObserver(LogObserverHolder *observer_holder);
+
+    void ResetDefaultLogObserver();
```
[the remark this answers is not in the document]
Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/diag/diag_log_observer.hpp:35]: They are, but it's internal/not exposed. Why are you exposing it?

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    void DefaultLogObserver(const LogMetaData &log_metadata, const LogBody &log_body, void *user_data) {
+        std::scoped_lock lk(g_log_observer_lock);
+
+        const char *color_type = nullptr;
+        /* Get the color according to the log severity. */
+        if (log_metadata.log_severity < LogSeverity_Count) {
+            color_type = g_log_severity_color_table[log_metadata.log_severity];
+        }
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:43]: line break after here.

File: libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_string.hpp [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+#include <vapours.hpp>
+
+namespace ams::diag::detail {
+
+    void PrintDebugString(const char *str, size_t len);
```
[the remark this answers is not in the document]
Comment (SciresM) [reply libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_string.hpp:21]: It does, but it's internal, and should probably not be exposed.

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+        auto has_module_name = false;
+        if (log_metadata.module_name != nullptr) {
+            has_module_name = strlen(log_metadata.module_name) > 1; /* Avoid '$' as module name. */
+        }
+
+        if (!has_module_name && !color_type) {
+            /* Just send the text log. */
+            PrintDebugString(log_body.log_text, log_body.log_text_length);
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:51]: return PrintDebugString

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+        std::scoped_lock lk(g_log_observer_lock);
+
+        const char *color_type = nullptr;
+        /* Get the color according to the log severity. */
+        if (log_metadata.log_severity < LogSeverity_Count) {
+            color_type = g_log_severity_color_table[log_metadata.log_severity];
+        }
+        auto has_module_name = false;
+        if (log_metadata.module_name != nullptr) {
+            has_module_name = strlen(log_metadata.module_name) > 1; /* Avoid '$' as module name. */
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:46]: Comments go on their own lines.

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 36]
```diff
@@ -0,0 +1,92 @@
...
+
+    void DefaultLogObserver(const LogMetaData &log_metadata, const LogBody &log_body, void *user_data) {
+        std::scoped_lock lk(g_log_observer_lock);
+
+        const char *color_type = nullptr;
+        /* Get the color according to the log severity. */
+        if (log_metadata.log_severity < LogSeverity_Count) {
+            color_type = g_log_severity_color_table[log_metadata.log_severity];
+        }
+        auto has_module_name = false;
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:44]: ...

const auto has_module_name = log_metadata.module_name != nullptr && std::strlen(log_metadata.module_name) > 1?

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 37]
```diff
@@ -0,0 +1,92 @@
...
+
+        if (!has_module_name && !color_type) {
+            /* Just send the text log. */
+            PrintDebugString(log_body.log_text, log_body.log_text_length);
+            return;
+        }
+
+        size_t used_buffer_space = 0;
+        if (has_module_name && log_body.log_is_head) {
+            auto fmt_prefix = (color_type != nullptr) ? color_type : "";
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:57]: const auto?

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+
+            auto module_name = log_metadata.module_name;
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:59]: what type is this? inappropriate auto usage.

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 39]
```diff
@@ -0,0 +1,92 @@
...
+            if (module_name[0] == '$') {
+                /* Skip logging this initial character if it's present. */
+                module_name++;
+            }
+
+            auto system_tick = os::GetSystemTick();
+            auto tick_ts = os::ConvertToTimeSpan(system_tick);
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:66]: ... const auto tick = os::ConvertToTimeSpan(os::GetSystemTick()); ?

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 40]
```diff
@@ -0,0 +1,92 @@
...
+            used_buffer_space += util::SNPrintf(g_default_log_observer_text_buffer, sizeof(g_default_log_observer_text_buffer), "%s%d:%02d:%02d.%03d [%-5s] ", fmt_prefix, tick_ts.GetHours(), tick_ts.GetMinutes(), tick_ts.GetSeconds(), tick_ts.GetMilliSeconds(), module_name);
+        }
+        else if (color_type) {
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:69]: invalid else format

Comment (SciresM) [reply libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:69]: color_type != nullptr

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 41]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        auto log_text_size = std::min(left_buffer_space, log_body.log_text_length);
+        if (log_body.log_text_length && (log_body.log_text[log_body.log_text_length - 1] == '\n')) {
+            /* If the last character is a newline character, skip it. */
+            log_text_size--;
+        }
+
+        /* If we used a color before, reset it now. */
+        auto fmt_prefix = (color_type != nullptr) ? "\x1B[0m" : "";
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:86]: Why is this a local variable, and not e.g. a constexpr variable which is ternary'd in the SNPrintf call?

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 42]
```diff
@@ -0,0 +1,92 @@
...
+
+            auto system_tick = os::GetSystemTick();
+            auto tick_ts = os::ConvertToTimeSpan(system_tick);
+            used_buffer_space += util::SNPrintf(g_default_log_observer_text_buffer, sizeof(g_default_log_observer_text_buffer), "%s%d:%02d:%02d.%03d [%-5s] ", fmt_prefix, tick_ts.GetHours(), tick_ts.GetMinutes(), tick_ts.GetSeconds(), tick_ts.GetMilliSeconds(), module_name);
+        }
+        else if (color_type) {
+            used_buffer_space += util::SNPrintf(g_default_log_observer_text_buffer, sizeof(g_default_log_observer_text_buffer), "%s", color_type);
+        }
+
+        auto left_buffer_space = sizeof(g_default_log_observer_text_buffer) - used_buffer_space;
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:73]: ...const auto max_buffer_size = sizeof(....) - used_buffer_space - (color_type != nullptr ? std::strlen(ResetColorType) - 1 : 0)

File: libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp [hunk 43]
```diff
@@ -0,0 +1,92 @@
...
+        if (color_type) {
+            /* Also subtract the size of the color formatting, if present. */
+            left_buffer_space -= strlen(color_type);
+        }
+
+        auto log_text_size = std::min(left_buffer_space, log_body.log_text_length);
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:79]: const auto, specify that it's std::min<size_t>

File: libraries/libstratosphere/source/diag/detail/diag_detail_log.cpp [hunk 44]
```diff
@@ -0,0 +1,128 @@
...
+        struct LogPrintData {
+            char *print_str_start;
+            char *print_str_end;
+            char *print_str_cur;
+            const LogMetaData &log_metadata;
+            bool is_head;
+            char buffer[40];
+        };
+
+        void PutCharacters(void *user_data, const char *chars, int chars_len) {
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/detail/diag_detail_log.cpp:31]: This needs comments, to not use auto everywhere, to use const where appropriate...

File: libraries/libstratosphere/source/fs/fs_sd_card.cpp [hunk 45]
```diff
@@ -43,6 +43,22 @@ namespace ams::fs {
...
+        if (this->notifier != nullptr) {
+            fsEventNotifierClose(static_cast<FsEventNotifier *>(this->notifier));
+        }
+    }
+
+    Result SdCardDetectionEventNotifier::GetEventHandle(os::SystemEvent *out_event, os::EventClearMode clear_mode) {
+        AMS_ABORT_UNLESS(this->notifier != nullptr);
+
+        ::Event event;
+        R_TRY(fsEventNotifierGetEventHandle(static_cast<FsEventNotifier *>(this->notifier), std::addressof(event), clear_mode == os::EventClearMode_AutoClear));
```
Comment (SciresM) [inline libraries/libstratosphere/source/fs/fs_sd_card.cpp:56]: This needs fixes, which I'll do.

File: libraries/libstratosphere/source/diag/diag_log_observer.cpp [hunk 46]
```diff
@@ -0,0 +1,89 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stratosphere.hpp>
+
+namespace ams::diag {
```
Comment (SciresM) [inline libraries/libstratosphere/source/diag/diag_log_observer.cpp:18]: This and the other log observer file should really be a single translation unit. I'll handle the merge.

File: libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp [hunk 47]
```diff
@@ -0,0 +1,100 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stratosphere.hpp>
+
+namespace ams::lm::detail {
+
+    namespace {
+
+        os::SdkMutex g_log_observer_lock;
+        u8 g_packet_transmitter_buffer[0x400];
```
Comment (SciresM) [inline libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp:23]: You only have a single buffer, N has a unit heap of 4, but I can handle this.

File: libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp [hunk 48]
```diff
@@ -0,0 +1,100 @@
...
+        bool LogPacketTransmitterFlushFunction(const void *log_data, size_t log_data_size) {
+            /* TODO: libnx bindings. */
+            return true;
+        }
+
+        constexpr const char DefaultInvalidString[] = "(Invalid UTF8 string)";
+
+    }
+
+    void LogManagerLogObserver(const diag::LogMetaData &log_metadata, const diag::LogBody &log_body, void *user_data) {
```
Comment (SciresM) [inline libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp:34]: This (and other code) creates lots of locals that are non-const and never used again. Better off with immediates.

**Review by SciresM — CHANGES_REQUESTED** [review]
Reviewed about half the files (no actual lm process code yet).

Mostly stylistic stuff. Do a refactor for the ::detail:: -> ::impl::, ::impl:: -> ::srv:: stuff mentioned before, then I'll do a pass over the lm process code.

**Pull request closed without merging**